File: rec_mem.js
function caminhoDasPedras(caminho, cansado = false, pos = 0, table = {}) {
	if (caminho[pos] == "0" || pos >= caminho.length) return 0;
	if (caminho[pos] == "m" && pos == caminho.length - 1) return 1;

	if (table[pos] && table[pos][cansado] != undefined) {
		return table[pos][cansado];
	}

	let res = 0;
	for (let i = 1; i <= 3 - cansado; i++) {
		res += caminhoDasPedras(caminho, i == 3, pos + i, table);
	}

	if (!table[pos]) table[pos] = {};

	table[pos][cansado] = res;

	return res;
}

const args = process.argv.slice(2);

if (args.length != 1) {
	console.error("Uso: node rec_mem.js <caminho>");
	process.exit(1);
}

console.log(caminhoDasPedras(args[0]));

File: src/bin/rec_mem.rs
use std::collections::HashMap;

fn caminho_das_pedras(
    caminho: &str,
    cansado: bool,
    table: &mut HashMap<(usize, bool), usize>,
) -> usize {
    match caminho.chars().next() {
        Some('0') | None => 0,
        Some('m') if caminho.len() == 1 => 1,
        _ => match table.get(&(caminho.len(), cansado)) {
            Some(res) => *res,
            None => {
                let res = (1..=caminho.len().min(3 - cansado as usize))
                    .map(|x| caminho_das_pedras(&caminho[x..], x == 3, table))
                    .sum();

                table.insert((caminho.len(), cansado), res);

                res
            }
        },
    }
}

fn main() {
    let args = std::env::args().collect::<Vec<_>>();

    if args.len() != 2 {
        println!("Uso: {} <caminho>", args[0]);
        return;
    }

    println!(
        "{}",
        caminho_das_pedras(&args[1], false, &mut HashMap::new())
    );
}
